cli/testflinger_cli/tests/test_cli.py

```diff
@@ -85,6 +89,61 @@ def test_submit(capsys, tmp_path, requests_mock):
     assert jobid in std.out
 
 
+def test_submit_with_attachments(tmp_path):
+    """Make sure jobs with attachments are submitted correctly"""
+
+    job_id = str(uuid.uuid1())
+    job_file = tmp_path / "test.json"
+    job_data = {
+        "queue": "fake",
+        "test_data": {
+            "attachments": [
+                {
+                    # include the submission JSON itself as a test attachment
+                    "local": str(job_file)
+                }
+            ]
+        },
+    }
+    job_file.write_text(json.dumps(job_data))
+
+    sys.argv = ["", "submit", str(job_file)]
+    tfcli = testflinger_cli.TestflingerCli()
+
+    with Mocker() as mocker:
+
+        # register responses for job and attachment submission endpoints
+        mock_response = {"job_id": job_id}
+        mocker.post(f"{URL}/v1/job", json=mock_response)
+        mocker.post(f"{URL}/v1/job/{job_id}/attachments")
+
+        # use cli to submit the job (processes `sys.argv` for arguments)
+        tfcli.submit()
+
+        # check the request history to confirm that:
+        # - there is a request to the job submission endpoint
+        # - there a request to the attachment submission endpoint
+        history = mocker.request_history
+        assert len(history) == 2
+        assert history[-2].path == "/v1/job"
+        assert history[-1].path == f"/v1/job/{job_id}/attachments"
+
+        # extract the binary file data from the request
+        # (`requests_mock` only provides access to the `PreparedRequest`)
+        match = re.search(b"\r\n\r\n(.+)\r\n--", history[-1].body, re.DOTALL)
```

Comment:
Is this the only way how to retrieve it? Seems fragilie (what if chunked encoding is used? etc etc...)

Reply:
This is a unit test to intercept (and check) the request sent by the Testflinger CLI. I'm afraid there's currently no way to retrieve the actual `Request` (as opposed to the binary `PreparedRequest`) and I couldn't come up with a better alternative. If there is a change in the way the CLI makes the request then the test might also need to change to reflect that.